Added tests for Group service

zanpizmoht opened the pull request:
Added tests for adding, deleting and editing groups, and adding devices to groups.

Signed-off-by: zanpizmoht <zan.pizmoht@comtrade.com>

Test scenarios for Groups are added in this PR.

**Related Issue**
_None_

**Description of the solution adopted**
Test scenarios that are added in this PR are checking functionalities regarding Groups. All scenarios are added to `GroupService.feature`. Scenarios are testing if Groups are correctly created, edited or deleted and if Devices can be added to Groups.

**Screenshots**
/

Coduz commented:
CQ submitted: [CQ 21395](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=21395)

LeoNerdoG commented on qa/integration/src/test/resources/features/authorization/GroupService.feature:
```diff
@@ -256,3 +256,1053 @@ Feature: Group Service tests
+      | integer | maxNumberChildEntities | 10    |
+    When I create the group with name "Group1" and description "Description 1234"
+    And I create a device with name "Device1"
+    And I add device to group
```
Change step name to: "And I add device "Device1" to group "Group1"

LeoNerdoG commented on qa/integration/src/test/resources/features/authorization/GroupService.feature:
```diff
@@ -256,3 +256,1053 @@ Feature: Group Service tests
+    And I create a device with name "Device1"
+    And I add device to group
+    And I search for a group named "Group1"
+    Then Device is in groups Assigned Devices
```
Change step name to: "Then Device "Device1" is in Assigned Devices of group "Group1" so we can use this more clearly in other scenarios as well.

LeoNerdoG commented on qa/integration/src/test/resources/features/authorization/GroupService.feature:
```diff
@@ -256,3 +256,1053 @@ Feature: Group Service tests
+    And No exception was thrown
+    And I logout
+
+  Scenario: Change Group With Description of Device
```
Change title to: "Moving Device From One Group With Description to Another

LeoNerdoG commented on qa/integration/src/test/resources/features/authorization/GroupService.feature:
```diff
@@ -256,3 +256,1053 @@ Feature: Group Service tests
+
+  Scenario: Change Group With Description of Device
+  Login as kapua-sys, go to groups, create a group.
+  Go to devices, create a device ("Device1") and add created group to it.
```
Change description to: "Go to devices, create a device ("Device1") and add it to the created group."

LeoNerdoG commented on qa/integration/src/test/resources/features/authorization/GroupService.feature:
```diff
@@ -256,3 +256,1053 @@ Feature: Group Service tests
+  Login as kapua-sys, go to groups, create a group.
+  Go to devices, create a device ("Device1") and add created group to it.
+  Check if created device is in Asigned Devices of the group.
+  Create another group and add it to "Device1"
```
Change description to: "Create another group and add device ("Device1") to it."

LeoNerdoG commented on qa/integration/src/test/resources/features/authorization/GroupService.feature:
```diff
@@ -256,3 +256,1053 @@ Feature: Group Service tests
+  Check if created device is in Asigned Devices of the group.
+  Create another group and add it to "Device1"
+  Check both groups Assigned Devices ("Device1" should only be in second group)
+  Kapua should throw Exception.
```
Change description to: "Kapua should throw Exception  when checking for device in "Group1"".

LeoNerdoG commented on qa/integration/src/test/resources/features/authorization/GroupService.feature:
```diff
@@ -256,3 +256,1053 @@ Feature: Group Service tests
+      | integer | maxNumberChildEntities | 10    |
+    When I create the group with name "Group1" and description "Description 123"
+    And I create a device with name "Device1"
+    And I add device to group
```
Change step name to: "And I add device "Device1" to group "Group1""

LeoNerdoG commented on qa/integration/src/test/resources/features/authorization/GroupService.feature:
```diff
@@ -256,3 +256,1053 @@ Feature: Group Service tests
+    And I create a device with name "Device1"
+    And I add device to group
+    When I search for a group named "Group1"
+    Then Device is in groups Assigned Devices
```
Change step name to: "Then Device "Device1" is in Assigned Devices of group "Group1"

LeoNerdoG commented on qa/integration/src/test/resources/features/authorization/GroupService.feature:
```diff
@@ -256,3 +256,1053 @@ Feature: Group Service tests
+    When I search for a group named "Group1"
+    Then Device is in groups Assigned Devices
+    When I create the group with name "Group2" and description "Description 123"
+    And I change devices group
```
Change step name to: "And I move device "Device1" from group "Group1" to group "Group2"

LeoNerdoG commented on qa/integration/src/test/resources/features/authorization/GroupService.feature:
```diff
@@ -256,3 +256,1053 @@ Feature: Group Service tests
+    When I create the group with name "Group2" and description "Description 123"
+    And I change devices group
+    And I search for a group named "Group2"
+    Then Device is in groups Assigned Devices
```
Change step name to: "Then Device "Device1" is in Assigned Devices of group "Group2"

LeoNerdoG commented on qa/integration/src/test/resources/features/authorization/GroupService.feature:
```diff
@@ -256,3 +256,1053 @@ Feature: Group Service tests
+    And I search for a group named "Group2"
+    Then Device is in groups Assigned Devices
+    When I search for a group named "Group1"
+    Then Device is not in groups Assigned Devices
```
Change step name to: "Then Device "Device1" is not in Assigned Devices of group "Group1""

LeoNerdoG commented on service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java:
```diff
@@ -2559,4 +2560,65 @@ public void iFindDeviceWithClientId(String deviceName) {
         assertEquals(device.getClientId(), deviceName);
     }
+
+    @And("^I add device to group$")
```
Change step name to: "I add device "Device1" to group "Group1""

LeoNerdoG commented on service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java:
```diff
@@ -2559,4 +2560,65 @@ public void iFindDeviceWithClientId(String deviceName) {
     }
+
+    @And("^I add device to group$")
+    public void iAddDeviceToGroup() throws Exception{
```
Change method name to: "public void iAddDeviceToGroup(String deviceName, String groupName) throws Exception{

LeoNerdoG commented on service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java:
```diff
@@ -2559,4 +2560,65 @@ public void iFindDeviceWithClientId(String deviceName) {
+
+    @And("^I add device to group$")
+    public void iAddDeviceToGroup() throws Exception{
+        Device device = (Device) stepData.get("Device");
```
Change method implementation, so it accepts the parameters that are given to it; do not use "stepData.get("Device")

LeoNerdoG commented on service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java:
```diff
@@ -2559,4 +2560,65 @@ public void iFindDeviceWithClientId(String deviceName) {
+    @And("^I add device to group$")
+    public void iAddDeviceToGroup() throws Exception{
+        Device device = (Device) stepData.get("Device");
+        Group group = (Group) stepData.get("Group");
```
Change method implementation, so it accepts the parameters that are given to it; do not use "stepData.get("Group")

LeoNerdoG commented on service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java:
```diff
@@ -2559,4 +2560,65 @@ public void iFindDeviceWithClientId(String deviceName) {
+        }
+    }
+
+    @When("^I change devices group to No Group$")
```
Change step name to: "Then Device "Device1" is removed from all groups".

LeoNerdoG commented on service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java:
```diff
@@ -2559,4 +2560,65 @@ public void iFindDeviceWithClientId(String deviceName) {
+    }
+
+    @When("^I change devices group to No Group$")
+    public void iChangeDevicesGroupToNoGroup() throws Exception{
```
Add parameter to the method: "public void iChangeDevicesGroupToNoGroup(String deviceName) throws Exception{"

LeoNerdoG commented on service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java:
```diff
@@ -2559,4 +2560,65 @@ public void iFindDeviceWithClientId(String deviceName) {
+
+    @When("^I change devices group to No Group$")
+    public void iChangeDevicesGroupToNoGroup() throws Exception{
+        Device device = (Device) stepData.get("Device");
```
Use "Device1" paramater and not "last" device.

LeoNerdoG commented on service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java:
```diff
@@ -2559,4 +2560,65 @@ public void iFindDeviceWithClientId(String deviceName) {
+        }
+    }
+
+    @And("^I create (\\d+) devices and add them to group$")
```
Change step name to: "@And("^I create (\\d+) devices and add them to group "Group1"$")

LeoNerdoG commented on service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java:
```diff
@@ -2559,4 +2560,65 @@ public void iFindDeviceWithClientId(String deviceName) {
+    }
+
+    @And("^I create (\\d+) devices and add them to group$")
+    public void iCreateDevicesAndAddThemToGroup(int numberOfDevices) throws Exception {
```
Change method implementation so it accepts group parameter: " public void iCreateDevicesAndAddThemToGroup(int numberOfDevices, String group) throws Exception {"

LeoNerdoG commented on service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java:
```diff
@@ -2559,4 +2560,65 @@ public void iFindDeviceWithClientId(String deviceName) {
+        }
+    }
+
+    @And("^Device has no group$")
```
This is not used anywhere?

LeoNerdoG commented on service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java:
```diff
@@ -2559,4 +2560,65 @@ public void iFindDeviceWithClientId(String deviceName) {
+        }
+    }
+
+    @And("^I update device$")
```
This is also not used here?

LeoNerdoG reviewed (changes requested):
Hi @zanpizmoht,
I have reviewed the PR and there are some changes that need to be made.

LeoNerdoG reviewed (changes requested):
Hi @zanpizmoht,
I have reviewed the PR and there are some changes that need to be made.

LeoNerdoG commented on qa/integration/src/test/resources/features/authorization/GroupService.feature:
```diff
@@ -256,3 +256,792 @@ Feature: Group Service tests
+    Then No exception was thrown
+    Then I logout
+
+  Scenario: Creating Unique Tag With Long Description
```
Please change this title to: "Creating Unique Group With Long Description" becasue we are creating groups here and not tags =)

LeoNerdoG reviewed (commented):
Hey, @zanpizmoht, there are only 2 minor things left to change. 
Please change them and push to PR. 
@Coduz this PR can be labeled as "QA Approved" label.

LeoNerdoG reviewed (approved):
Hi @zanpizmoht,
I have checked all the changes and everything is OK now. 
@coduz, can you please restart the failing tests and merge when CQ passes?

Coduz commented:
[CQ Approved!](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=21395#c6) 🎉

Coduz commented on qa/integration/src/test/resources/features/user/UserPermissionI9n.feature:
```diff
@@ -1441,8 +1441,13 @@ Feature: User Permission tests
+=======
+    And I create a tag with name "test_tag"
+    And I create a group with name "test_group"
+>>>>>>> Added tests for Group service
```
Please remove the merge meta data

zanpizmoht committed:
Added tests for Group service
Added tests for adding, deleting and editing groups, and adding devices to groups.

Signed-off-by: zanpizmoht <zan.pizmoht@comtrade.com>

zanpizmoht commented:
Hi @Coduz,
I fixed the errors and I saw Travis just passed so this PR is ready to be merged.
Regards, Zan
